### PR TITLE
Plans 2023: header-price refactors and unit tests

### PR DESCRIPTION
**Pull request by chriskmnds**
Related to https://github.com/Automattic/wp-calypso/issues/74261

## Proposed Changes

This is a follow-up to https://github.com/Automattic/wp-calypso/pull/74375 (which we have put on hold) to compile the relevant refactors that we'd like to merge into `trunk`, along with some unit/UI tests for the header-price component (should also cover usage of the new hook). We are introducing a `use-plan-prices` hook under `/plans`, something that we'll be able to repurpose into other refactors and work pricing-related.

Changes here should allow us to quickly pick up developments with  https://github.com/Automattic/wp-calypso/issues/74261 once we have a clearer idea of cost-savings.

## Testing Instructions

* Code review
* Unit tests
* Go to `/plans/[SITE]` with a variety of sites (e.g. on free plan, on paid plan, with pro-rated credits, etc.) and confirm prices are displayed as expected (with cross-out prices when discounts exist)
* Confirm with different currencies e.g. IRN to ensure those discounts are also visible in the prices

p.s. This is a foundation PR that will also be tested more thoroughly in a follow-up PR that will be based on this one to update the price displays.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/plan-features-2023-grid/header-price.tsx
Comment (aneeshd16) [inline client/my-sites/plan-features-2023-grid/header-price.tsx]: Since we are already checking for `shouldShowDiscountedPrice` in the component, maybe the `discountedPricing` object is redundant? This could be:
```suggestion
							rawPrice={ planPrices.rawPrice }
```

File: client/my-sites/plan-features-2023-grid/header-price.tsx
Comment (aneeshd16) [inline client/my-sites/plan-features-2023-grid/header-price.tsx]: Same as above,
```
rawPrice={ planPrices.planDiscountedRawPrice || planPrices.discountedRawPrice }
```

File: client/my-sites/plans/hooks/use-plan-prices.ts
Comment (aneeshd16) [inline client/my-sites/plans/hooks/use-plan-prices.ts]: It's not clear what this param does. Can you please explain or add a comment?

**Review by aneeshd16 — APPROVED** [review]
Tests out! I added a non-blocking code comment below.

File: client/my-sites/plan-features-2023-grid/header-price.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/header-price.tsx]: Good call, @aneeshd16. Yes, it is redundant. I think the intention with keeping it separate was to more easily introduce conditioning needed for the cost savings (where the refactors came from). but we'll cross that line when we get to it.

File: client/my-sites/plans/hooks/use-plan-prices.ts
Comment (chriskmnds) [reply client/my-sites/plans/hooks/use-plan-prices.ts]: Good point. It's the boolean passed to the selectors below, akin to `isMonthly`. They in turn fetch full-term or monthly-projected prices from what I can tell.

I prefer `monthly` (in contrast to `isMonthly`), I believe mainly due to not having a reference to something/"it" that is monthly or not. It's just a computation projection. The real semantic I guess would have been "projectToMonthly".

I see `getPlanDiscountedRawPrice` is using `isMonthly`, which I would have also considered for refactoring in similar vein 🤔

Comment (chriskmnds) [reply client/my-sites/plans/hooks/use-plan-prices.ts]: Let me know if you prefer renaming to something and will happily make the change
I'm actually inclined to go with `projectToMonthly` now 🤔

File: client/my-sites/plan-features-2023-grid/header-price.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/header-price.tsx]: Updated. Thanks for the great questions/suggestions.
Let's do another round of testing maybe jic 😬

File: client/state/plans/selectors/get-discounted-raw-price.js
Comment (sirbrillig) [inline client/state/plans/selectors/get-discounted-raw-price.js:19]: This made me realize that this function was missing a failsafe for missing products so I'm adding one here: https://github.com/Automattic/wp-calypso/pull/75140

File: client/my-sites/plans/hooks/use-plan-prices.ts
Comment (sirbrillig) [reply client/my-sites/plans/hooks/use-plan-prices.ts]: > I see getPlanDiscountedRawPrice is using isMonthly, which I would have also considered for refactoring in similar vein

I saw the same thing but before I read this comment. In https://github.com/Automattic/wp-calypso/pull/75140 I'm suggesting changing it to `returnMonthly`; what do you think? I also left a comment to explain what it does (and the risks of relying on it).

Comment (chriskmnds) [reply client/my-sites/plans/hooks/use-plan-prices.ts]: Thanks, @sirbrillig. `returnMonthly` works for me. I can use the same here.

Comment (aneeshd16) [reply client/my-sites/plans/hooks/use-plan-prices.ts]: `returnMonthly` sounds good to me!

**Review by aneeshd16 — APPROVED** [review]
Re-tested, LGTM after making the `returnMonthly` change. 👍

**Comment by chriskmnds** [comment]
Thank y'all for testing - will ship once the checks pass.

@sirbrillig you'll need to rebase https://github.com/Automattic/wp-calypso/pull/75140 to update the call from use-plan-prices once merged.

**Pull request merged**
